src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php

```diff
@@ -112,21 +112,24 @@ public function onKernelResponse(FilterResponseEvent $event)
     protected function injectToolbar(Response $response, Request $request)
     {
         $content = $response->getContent();
-        $pos = strripos($content, '</body>');
-
-        if (false !== $pos) {
-            $toolbar = "\n".str_replace("\n", '', $this->twig->render(
-                '@WebProfiler/Profiler/toolbar_js.html.twig',
-                array(
-                    'position' => $this->position,
-                    'excluded_ajax_paths' => $this->excludedAjaxPaths,
-                    'token' => $response->headers->get('X-Debug-Token'),
-                    'request' => $request,
-                )
-            ))."\n";
-            $content = substr($content, 0, $pos).$toolbar.substr($content, $pos);
-            $response->setContent($content);
+        $injectToolbarAt = strripos($content, '</body>') ?: strripos($content, '</html>');
+        $canInjectToolbar = $injectToolbarAt || stripos($content, '<title>');
```

Comment:
you can have a normal HTTP request returning partial HTML (generally done through XHR though, but detecting XHR requests cannot be done. ``isXMLHttpRequest`` detects a special header added by jQuery and a few other major frameworks, but it may not be there when using the native APIs directly).

Another case of partial is the response to an ESI request (which is **not** a subrequest, especially when the ESI is resolved outside Symfony)